pulsar-client-cpp/tests/KeyValueImplTest.cc

```diff
@@ -0,0 +1,116 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#include <gtest/gtest.h>
+#include <KeyValueImpl.h>
+
+using namespace pulsar;
+
+TEST(KeyValueTest, testEncodeAndDeCode) {
+    const std::string keyContent = "keyContent";
+    const std::string valueContent = "valueContent";
+
+    {
+        // test inline encode
+        KeyValueImpl keyValue((std::string(keyContent)), std::string(valueContent));
```

Comment:
I found there are many places `KeyValueTest` you wrote something like

```
std::string keyContent = "key";
std::string valueContent = "value";
KeyValueImpl keyValue(std::string(keyContent), std::string(valueContent));
```

The code looks complicated and redundant. If you have two pre-defined `std::string` objects like:

```java
std::string keyContent = "key";
std::string valueContent = "value";
```

You can just use `std::move` to move these strings into the arguments of `KeyValueImpl` constructor like:

```c++
KeyValueImpl keyValue(std::move(keyContent), std::move(valueContent));
```

Since const objects cannot be moved, you should remove the `const` qualifier on `keyContent` and `valueContent`.

If you want to reuse two strings, you can declare `const char*` variables and construct the `std::string` implicitly.

Here is an example that shows how I rewrote the `testValueIsEmpty`:

```c++
TEST(KeyValueTest, testValueIsEmpty) {
    const char* keyContent = "keyContent";
    const char* valueContent = "";

    {
        // test inline encode
        KeyValueImpl keyValue(keyContent, valueContent);
        const SharedBuffer content = keyValue.getContent(KeyValueEncodingType::INLINE);
        ASSERT_EQ(content.readableBytes(), 8 + strlen(keyContent) + strlen(valueContent));

        // test inline decode
        KeyValueImpl deCodeKeyValue(content.data(), content.readableBytes(), KeyValueEncodingType::INLINE);
        const SharedBuffer deCodeContent = keyValue.getContent(KeyValueEncodingType::INLINE);
        ASSERT_EQ(deCodeKeyValue.getKey(), keyContent);
        ASSERT_EQ(deCodeKeyValue.getValueAsString(), valueContent);
        ASSERT_NE(std::string(deCodeContent.data(), deCodeContent.readableBytes()), valueContent);
    }

    {
        // test separated type
        KeyValueImpl sepKeyValue(keyContent, valueContent);
        const SharedBuffer content = sepKeyValue.getContent(KeyValueEncodingType::SEPARATED);
        ASSERT_EQ(sepKeyValue.getKey(), keyContent);
        ASSERT_EQ(sepKeyValue.getValueAsString(), valueContent);
        ASSERT_EQ(std::string(content.data(), content.readableBytes()), valueContent);
    }
}
```

Reply:
Thanks, learned, this is a good idea, I changed it.